File: day10/day10.py
#!/usr/bin/env python3

from collections import deque
import sys

def part1(nums):
    d1, d3 = 1, 1
    for i, value in enumerate(nums[1:], 1):
        if value - 1 == nums[i-1]:
            d1 += 1
        else:
            for j in range(1, 4):
                if i - j > 0 and value - 3 == nums[i-j]:
                    d3 += 1
                    break
    return d1 * d3

def part2(nums):
    q = deque([0] * 3)
    count = 1
    i = len(nums)-1
    while i > 0:
        for j in range(3):
            if i > j and nums[i] - nums[i-j-1] <= 3:
                q[j] += count

        count = q.popleft()
        q.append(0)
        i -= 1

    return count

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    nums = sorted(map(int, txt.splitlines()))

    print("Part1:", part1(nums))

    nums = [0] + nums + [nums[-1]+3]
    print("Part2:", part2(nums))

File: day10/day10.cpp
#include <algorithm>
#include <fstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

static int joltsmul(const vector<int>& jolts)
{
	int d1 = 1;
	int d3 = 1;
	for (size_t i = 1; i < jolts.size(); i++)
	{
		// check for value-1
		if (jolts[i] - 1 == jolts[i-1])
		{
			d1++;
		}
		else
		{
			// check for value-3
			for (size_t j = 1; j < 4; j++)
			{
				if (i >= j && jolts[i] - 3 == jolts[i - j])
				{
					d3++;
					break;
				}
			}
		}
	}
	return d1 * d3;
}

static uint64_t solutions(const vector<int>& jolts)
{
	// array with the last 4 solutions, index 0 is the last and is
	// set to 1 initially because for the last element we have
	// exactly 1 choice. The fifth element is a convenience one
	// which never changes from zero.
	uint64_t buf[5];
	fill(buf+1, buf+5, 0);
	buf[0] = 1;

	// NOTE: we stop at 1 because i - 1 == 0 otherwise we make an
	// empty inner loop and we overwrite buf[0]
	size_t i = jolts.size();
	while (i-->1)
	{
		for (size_t j = 1; j < 4; j++)
		{
			if (i >= j && jolts[i] - jolts[i-j] <= 3)
			{
				buf[j] += buf[0];
			}
		}

		// NOTE: shift the values left by one position and a
		// zero element is placed at index 3 because element
		// at index 4 is always zero.
		copy(buf+1, buf+5, buf);
	}
	return buf[0];
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<int> jolts;
	int value;
	while (input >> value)
	{
		jolts.push_back(value);
	}
	input.close();

	sort(jolts.begin(), jolts.end());

	fmt::print("Part1: {}\n", joltsmul(jolts));

	jolts.insert(jolts.begin(), 0);
	jolts.push_back(jolts.back()+3);
	fmt::print("Part2: {}\n", solutions(jolts));

	return 0;
}
